README.md
Aurora
======

[![go.dev reference](https://img.shields.io/badge/go.dev-reference-007d9c?logo=go&logoColor=white)](https://pkg.go.dev/github.com/logrusorgru/aurora/v3?tab=doc)
[![Unlicense](https://img.shields.io/badge/license-unlicense-blue.svg)](http://unlicense.org/)
[![Build Status](https://github.com/logrusorgru/aurora/workflows/build/badge.svg)](https://github.com/logrusorgru/aurora/actions?workflow=build)
[![Coverage Status](https://coveralls.io/repos/github/logrusorgru/aurora/badge.svg?branch=master)](https://coveralls.io/github/logrusorgru/aurora?branch=master)
[![GoReportCard](https://goreportcard.com/badge/logrusorgru/aurora)](https://goreportcard.com/report/logrusorgru/aurora)

Ultimate ANSI colors for Golang. The package supports Printf/Sprintf etc.


![aurora logo](https://github.com/logrusorgru/aurora/blob/master/gopher_aurora.png)

# TOC

- [Installation](#installation)
- [Usage](#usage)
  + [Simple](#simple)
  + [Printf](#printf)
  + [aurora.Sprintf](#aurorasprintf)
  + [Enable/Disable colors](#enabledisable-colors)
  + [Hyperlinks, default colorizer, and configurations](#hyperlinks-default-colorizer-and-configurations)
- [Chains](#chains)
- [Colorize](#colorize)
- [Grayscale](#grayscale)
- [8-bit colors](#8-bit-colors)
- [Supported Colors & Formats](#supported-colors--formats)
  + [All colors](#all-colors)
  + [Standard and bright colors](#standard-and-bright-colors)
  + [Formats are likely supported](#formats-are-likely-supported)
  + [Formats are likely unsupported](#formats-are-likely-unsupported)
- [Limitations](#limitations)
  + [Windows](#windows)
  + [TTY](#tty)
- [Licensing](#licensing)

# Installation

##### Version 1.x

Using gopkg.in.

```
go get -u gopkg.in/logrusorgru/aurora.v1
```

##### Version 2.x

```
go get -u github.com/logrusorgru/aurora
```

##### Go modules support, version v3+

Get
```
go get -u github.com/logrusorgru/aurora/v3
```

The v3 was introduced to support `go.mod` and leave previous import paths as is.
Currently, there is no changes between them (excluding the importpath's /v3 tail).

##### The latest version

```
go get -u github.com/logrusorgru/aurora/v4
```

With hyperlinks.

# Test

```
go test -cover -race github.com/logrusorgru/aurora/v4
```

Replace the import path with your, if it's different.

# Usage

### Simple

```go
package main

import (
	"fmt"

	"github.com/logrusorgru/aurora/v4"
)

func main() {
	fmt.Println("Hello,", aurora.Magenta("Aurora"))
	fmt.Println(aurora.Bold(aurora.Cyan("Cya!")))
}

```

![simple png](https://github.com/logrusorgru/aurora/blob/master/simple.png)

### Printf

```go
package main

import (
	"fmt"

	"github.com/logrusorgru/aurora/v4"
)

func main() {
	fmt.Printf("Got it %d times\n", aurora.Green(1240))
	fmt.Printf("PI is %+1.2e\n", aurora.Cyan(3.14))
}

```

![printf png](https://github.com/logrusorgru/aurora/blob/master/printf.png)

### aurora.Sprintf

```go
package main

import (
	"fmt"

	"github.com/logrusorgru/aurora/v4"
)

func main() {
	fmt.Println(aurora.Sprintf(aurora.Magenta("Got it %d times"), aurora.Green(1240)))
}

```

![sprintf png](https://github.com/logrusorgru/aurora/blob/master/sprintf.png)

### Enable/Disable colors

```go
package main

import (
	"fmt"
	"flag"

	"github.com/logrusorgru/aurora/v4"
)

// colorizer
var au *aurora.Aurora

var colors = flag.Bool("colors", false, "enable or disable colors")

func init() {
	flag.Parse()
	au = aurora.New(WithColors(*colors))
}

func main() {
	// use colorizer
	fmt.Println(au.Green("Hello"))
}

```
Without flags:
![disable png](https://github.com/logrusorgru/aurora/blob/master/disable.png)

With `-colors` flag:
![enable png](https://github.com/logrusorgru/aurora/blob/master/enable.png)

### Hyperlinks, default colorizer, and configurations

[Hyperlinks feature description](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda).

Add a red hyperlinks with text "Example" that is referencing to
http://example.com.

```go
package main

import (
	"flag"
	"fmt"

	"github.com/logrusorgru/aurora/v4"
)

func main() {
	var conf = aurora.NewConfig()
	conf.AddFlags(flag.CommandLine, "prefix.")
	flag.Parse()

	aurora.DefaultColorizer = aurora.New(conf.Options()...) // set global

	fmt.Println(aurora.Red("Example").Hyperlink("http://example.com/"))
}
```
Depending flags:
![depending flags png](https://github.com/logrusorgru/aurora/blob/master/aurora_hyperlinks_flags.png)
![depending flags gif](https://github.com/logrusorgru/aurora/blob/master/aurora_hyperlinks.gif)

# Chains

The following samples are equal

```go
x := aurora.BgMagenta(aurora.Bold(aurora.Red("x")))
```

```go
x := aurora.Red("x").Bold().BgMagenta()
```

The second is more readable

# Colorize

There is `Colorize` function that allows to choose some colors and
format from a side

```go

func getColors() Color {
	// some stuff that returns appropriate colors and format
}

// [...]

func main() {
	fmt.Println(aurora.Colorize("Greeting", getColors()))
}

```
Less complicated example

```go
x := aurora.Colorize("Greeting", GreenFg|GrayBg|BoldFm)
```

Unlike other color functions and methods (such as Red/BgBlue etc)
a `Colorize` clears previous colors

```go
x := aurora.Red("x").Colorize(BgGreen) // will be with green background only
```

# Grayscale

```go
fmt.Println("  ",
	aurora.Gray(1-1, " 00-23 ").BgGray(24-1),
	aurora.Gray(4-1, " 03-19 ").BgGray(20-1),
	aurora.Gray(8-1, " 07-15 ").BgGray(16-1),
	aurora.Gray(12-1, " 11-11 ").BgGray(12-1),
	aurora.Gray(16-1, " 15-07 ").BgGray(8-1),
	aurora.Gray(20-1, " 19-03 ").BgGray(4-1),
	aurora.Gray(24-1, " 23-00 ").BgGray(1-1),
)
```

![grayscale png](https://github.com/logrusorgru/aurora/blob/master/aurora_grayscale.png)  

# 8-bit colors

Methods `Index` and `BgIndex` implements 8-bit colors.

| Index/BgIndex  |    Meaning      | Foreground | Background |
| -------------- | --------------- | ---------- | ---------- |
|      0-  7     | standard colors |   30- 37   |   40- 47   |
|      8- 15     | bright colors   |   90- 97   |  100-107   |
|     16-231     | 216 colors      |   38;5;n   |   48;5;n   |
|    232-255     | 24 grayscale    |   38;5;n   |   48;5;n   |

Example

```go
package main

import (
	"fmt"
	"github.com/logrusorgru/aurora"
)

func main() {
	for i := uint8(16); i <= 231; i++ {
		fmt.Println(i, aurora.Index(i, "pew-pew"), aurora.BgIndex(i, "pew-pew"))
	}
}
```

# Supported colors & formats

- formats
  + bold (1)
  + faint (2)
  + doubly-underline (21)
  + fraktur (20)
  + italic (3)
  + underline (4)
  + slow blink (5)
  + rapid blink (6)
  + reverse video (7)
  + conceal (8)
  + crossed out (9)
  + framed (51)
  + encircled (52)
  + overlined (53)
- background and foreground colors, including bright
  + black
  + red
  + green
  + yellow (brown)
  + blue
  + magenta
  + cyan
  + white
  + 24 grayscale colors
  + 216 8-bit colors

### All colors

![linux png](https://github.com/logrusorgru/aurora/blob/master/aurora_colors_black.png)  
![white png](https://github.com/logrusorgru/aurora/blob/master/aurora_colors_white.png)  

### Standard and bright colors

![linux black standard png](https://github.com/logrusorgru/aurora/blob/master/aurora_black_standard.png)
![linux white standard png](https://github.com/logrusorgru/aurora/blob/master/aurora_white_standard.png)

### Formats are likely supported

![formats supported gif](https://github.com/logrusorgru/aurora/blob/master/aurora_formats.gif)

### Formats are likely unsupported

![formats rarely supported png](https://github.com/logrusorgru/aurora/blob/master/aurora_rarely_supported.png)

# Limitations

There is no way to represent `%T` and `%p` with colors using
a standard approach

```go
package main

import (
	"fmt"

	"github.com/logrusorgru/aurora"
)

func main() {
	var (
		r = aurora.Red("red")
		i int
	)
	fmt.Printf("%T %p\n", r, aurora.Green(&i))
}
```

Output will be without colors

```
aurora.value %!p(aurora.value={0xc42000a310 768 0})
```

The obvious workaround is `Red(fmt.Sprintf("%T", some))`

### Windows

The Aurora provides ANSI colors only, so there is no support for Windows. That said, there are workarounds available.
Check out these comments to learn more:

- [Using go-colorable](https://github.com/logrusorgru/aurora/issues/2#issuecomment-299014211).
- [Using registry for Windows 10](https://github.com/logrusorgru/aurora/issues/10#issue-476361247).

### TTY

The Aurora has no internal TTY detectors by design. Take a look
 [this comment](https://github.com/logrusorgru/aurora/issues/2#issuecomment-299030108) if you want turn

CHANGELOG.md
Changes
=======

---
14:15:14
Thursday, October 8, 2022

New version v4. Breaking changes. [Migrate from v3](#migrate-from-v3).

- Drop deprecated `Bleach` methods. Use `Reset` or `Clear` instead.
- Drop deprecated `Brows` and `BgBrows` methods, and `BrownFg` and `BrownBg`
  colors. Use `Yellow` variants instead.
- Instead of `Aurora` interface introduced `Aurora` structure.
- Instead of `Value` interface introduced `Value` structure.
- Implemented [hyperlinks feature](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda).
  - added `Hyperlink` method
  - added `HyperlinkTarget` and `HyperlinkParams` methods
  - added `HyperlinkParam` type, `IsValidHyperlinkTarget`,
    `IsValidHyperlinkParam`, `HyperlinkID`, `HyperlinkEscape`,
    `HyperlinkUnescape` helper functions.
- Introduced `Config` and related functions and methods, such as `NewConfig`,
  `WithColors`, `WithHyperlinks`.
- Removed `NewAurora` function, new function `New` introduced.
- Introduced global `DefaultColorizer` that used by package root methods.
- `Sprintf` method now belongs to a colorizer and depends on its configurations.
  For package root `Sprintf` it's the `DefaultColorizer`.

Performance for all methods is almost the same. But for color- and
format-methods `aurora` now takes less allocations. But, unfortunately, for
`Sprintf` it takes more allocations.

###### Migrate from v3

1. Use `Reset` or `Clear` instead of `Bleach`.
2. Use `Yellow` instead of `Brows`.
3. Use `BgYellow` instead of `BgBrown`.
4. Use `YellowFg` instead of `BrownFg`.
5. Use `YellowBg` instead of `BrownBg`.
6. Use `New` instead of `NewAurora`.
7. Use `New(WithColors(false))` to disable colors.
8. Use `New(WithHyperlinks(false))` to disable hyperlinks.

---
16:05:02
Thursday, July 2, 2020

Change license from the WTFPL to the Unlicense due to pkg.go.dev restriction.

---
15:39:40
Wednesday, April 17, 2019

- Bright background and foreground colors
- 8-bit indexed colors `Index`, `BgIndex`
- 24 grayscale colors `Gray`, `BgGray`
- `Yellow` and `BgYellow` methods, mark Brow and BgBrown as deprecated
  Following specifications, correct name of the colors are yellow, but
  by historical reason they are called brown. Both, the `Yellow` and the
  `Brown` methods (including `Bg+`) represents the same colors. The Brown
  are leaved for backward compatibility until Go modules production release.
- Additional formats
  + `Faint` that is opposite to the `Bold`
  + `DoublyUnderline`
  + `Fraktur`
  + `Italic`
  + `Underline`
  + `SlowBlink` with `Blink` alias
  + `RapidBlink`
  + `Reverse` that is alias for the `Inverse`
  + `Conceal` with `Hidden` alias
  + `CrossedOut` with `StrikeThrough` alias
  + `Framed`
  + `Encircled`
  + `Overlined`
- Add AUTHORS.md file and change all copyright notices.
- `Reset` method to create clear value. `Reset` method that replaces
  `Bleach` method. The `Bleach` method was marked as deprecated.

---

14:25:49
Friday, August 18, 2017

- LICENSE.md changed to LICENSE
- fix email in README.md
- add "no warranty" to README.md
- set proper copyright date

---

16:59:28
Tuesday, November 8, 2016

- Rid out off sync.Pool
- Little optimizations (very little)
- Improved benchmarks

---


aurora.go
//
// Copyright (c) 2016-2022 The Aurora Authors. All rights reserved.
// This program is free software. It comes without any warranty,
// to the extent permitted by applicable law. You can redistribute
// it and/or modify it under the terms of the Unlicense. See LICENSE
// file for more details or see below.
//

//
// This is free and unencumbered software released into the public domain.
//
// Anyone is free to copy, modify, publish, use, compile, sell, or
// distribute this software, either in source code form or as a compiled
// binary, for any purpose, commercial or non-commercial, and by any
// means.
//
// In jurisdictions that recognize copyright laws, the author or authors
// of this software dedicate any and all copyright interest in the
// software to the public domain. We make this dedication for the benefit
// of the public at large and to the detriment of our heirs and
// successors. We intend this dedication to be an overt act of
// relinquishment in perpetuity of all present and future rights to this
// software under copyright law.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
// IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
// OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
// OTHER DEALINGS IN THE SOFTWARE.
//
// For more information, please refer to <http://unlicense.org/>
//

// Package aurora implements ANSI-colors
package aurora

type Aurora struct {
	conf Config
	cc   colorConfig
}

// New returns new colorizer by given Options.
func New(opts ...Option) (a *Aurora) {
	a = new(Aurora)
	a.conf = NewConfig()        // set defaults
	a.conf.Apply(opts...)       // apply options
	a.cc = a.conf.colorConfig() // keep the short hand
	return
}

// Config of the colorizer. It returns copy of the configurations.
func (a *Aurora) Config() Config {
	return a.conf
}

// Reset wraps given argument returning Value without formats, colors and links.
func (a *Aurora) Reset(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Reset()
	}
	return Value{
		cc:    a.cc,
		value: arg,
	}
}

// Clear wraps given argument returning Value without formats and colors. But
// preserving links.
func (a *Aurora) Clear(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Clear()
	}
	return Value{
		cc:    a.cc,
		value: arg,
	}
}

// Formats
//
// Bold or increased intensity (1).
func (a *Aurora) Bold(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Bold()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Bold()),
		value: arg,
	}
}

// Faint, decreased intensity (2).
func (a *Aurora) Faint(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Faint()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Faint()),
		value: arg,
	}
}

// DoublyUnderline or Bold off, double-underline per ECMA-48 (21).
func (a *Aurora) DoublyUnderline(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.DoublyUnderline()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).DoublyUnderline()),
		value: arg,
	}
}

// Fraktur, rarely supported (20).
func (a *Aurora) Fraktur(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Fraktur()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Fraktur()),
		value: arg,
	}
}

// Italic, not widely supported, sometimes treated as inverse (3).
func (a *Aurora) Italic(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Italic()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Italic()),
		value: arg,
	}
}

// Underline (4).
func (a *Aurora) Underline(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Underline()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Underline()),
		value: arg,
	}
}

// SlowBlink, blinking less than 150 per minute (5).
func (a *Aurora) SlowBlink(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.SlowBlink()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).SlowBlink()),
		value: arg,
	}
}

// RapidBlink, blinking 150+ per minute, not widely supported (6).
func (a *Aurora) RapidBlink(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.RapidBlink()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).RapidBlink()),
		value: arg,
	}
}

// Blink is alias for the SlowBlink.
func (a *Aurora) Blink(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Blink()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Blink()),
		value: arg,
	}
}

// Reverse video, swap foreground and background colors (7).
func (a *Aurora) Reverse(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Reverse()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Reverse()),
		value: arg,
	}
}

// Inverse is alias for the Reverse
func (a *Aurora) Inverse(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Inverse()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Inverse()),
		value: arg,
	}
}

// Conceal, hidden, not widely supported (8).
func (a *Aurora) Conceal(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Conceal()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Conceal()),
		value: arg,
	}
}

// Hidden is alias for the Conceal.
func (a *Aurora) Hidden(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Hidden()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Hidden()),
		value: arg,
	}
}

// CrossedOut, characters legible, but marked for deletion (9).
func (a *Aurora) CrossedOut(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.CrossedOut()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).CrossedOut()),
		value: arg,
	}
}

// StrikeThrough is alias for the CrossedOut.
func (a *Aurora) StrikeThrough(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.StrikeThrough()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).StrikeThrough()),
		value: arg,
	}
}

// Framed (51).
func (a *Aurora) Framed(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Framed()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Framed()),
		value: arg,
	}
}

// Encircled (52).
func (a *Aurora) Encircled(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Encircled()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Encircled()),
		value: arg,
	}
}

// Overlined (53).
func (a *Aurora) Overlined(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Overlined()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Overlined()),
		value: arg,
	}
}

// Foreground colors
//
// Black foreground color (30).
func (a *Aurora) Black(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Black()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Black()),
		value: arg,
	}
}

// Red foreground color (31).
func (a *Aurora) Red(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Red()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Red()),
		value: arg,
	}
}

// Green foreground color (32).
func (a *Aurora) Green(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Green()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Green()),
		value: arg,
	}
}

// Yellow foreground color (33).
func (a *Aurora) Yellow(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Yellow()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Yellow()),
		value: arg,
	}
}

// Blue foreground color (34).
func (a *Aurora) Blue(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Blue()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Blue()),
		value: arg,
	}
}

// Magenta foreground color (35).
func (a *Aurora) Magenta(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Magenta()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Magenta()),
		value: arg,
	}
}

// Cyan foreground color (36).
func (a *Aurora) Cyan(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Cyan()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Cyan()),
		value: arg,
	}
}

// White foreground color (37).
func (a *Aurora) White(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.White()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).White()),
		value: arg,
	}
}

// Bright foreground colors.
//
// BrightBlack foreground color (90).
func (a *Aurora) BrightBlack(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BrightBlack()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BrightBlack()),
		value: arg,
	}
}

// BrightRed foreground color (91).
func (a *Aurora) BrightRed(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BrightRed()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BrightRed()),
		value: arg,
	}
}

// BrightGreen foreground color (92).
func (a *Aurora) BrightGreen(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BrightGreen()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BrightGreen()),
		value: arg,
	}
}

// BrightYellow foreground color (93).
func (a *Aurora) BrightYellow(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BrightYellow()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BrightYellow()),
		value: arg,
	}
}

// BrightBlue foreground color (94).
func (a *Aurora) BrightBlue(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BrightBlue()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BrightBlue()),
		value: arg,
	}
}

// BrightMagenta foreground color (95).
func (a *Aurora) BrightMagenta(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BrightMagenta()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BrightMagenta()),
		value: arg,
	}
}

// BrightCyan foreground color (96).
func (a *Aurora) BrightCyan(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BrightCyan()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BrightCyan()),
		value: arg,
	}
}

// BrightWhite foreground color (97).
func (a *Aurora) BrightWhite(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BrightWhite()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BrightWhite()),
		value: arg,
	}
}

// Other colors.
//
// Index of pre-defined 8-bit foreground color from 0 to 255 (38;5;n).
//
//	  0-  7:  standard colors (as in ESC [ 30–37 m)
//	  8- 15:  high intensity colors (as in ESC [ 90–97 m)
//	 16-231:  6 × 6 × 6 cube (216 colors): 16 + 36 × r + 6 × g + b (0 ≤ r, g, b ≤ 5)
//	232-255:  grayscale from black to white in 24 steps
func (a *Aurora) Index(n ColorIndex, arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Index(n)
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Index(n)),
		value: arg,
	}
}

// Gray from 0 to 23.
func (a *Aurora) Gray(n GrayIndex, arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.Gray(n)
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).Gray(n)),
		value: arg,
	}
}

// Background colors.
//
// BgBlack background color (40).
func (a *Aurora) BgBlack(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgBlack()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgBlack()),
		value: arg,
	}
}

// BgRed background color (41).
func (a *Aurora) BgRed(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgRed()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgRed()),
		value: arg,
	}
}

// BgGreen background color (42).
func (a *Aurora) BgGreen(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgGreen()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgGreen()),
		value: arg,
	}
}

// BgYellow background color (43).
func (a *Aurora) BgYellow(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgYellow()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgYellow()),
		value: arg,
	}
}

// BgBlue background color (44).
func (a *Aurora) BgBlue(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgBlue()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgBlue()),
		value: arg,
	}
}

// BgMagenta background color (45).
func (a *Aurora) BgMagenta(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgMagenta()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgMagenta()),
		value: arg,
	}
}

// BgCyan background color (46).
func (a *Aurora) BgCyan(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgCyan()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgCyan()),
		value: arg,
	}
}

// BgWhite background color (47).
func (a *Aurora) BgWhite(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgWhite()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgWhite()),
		value: arg,
	}
}

// Bright background colors.
//
// BgBrightBlack background color (100).
func (a *Aurora) BgBrightBlack(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgBrightBlack()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgBrightBlack()),
		value: arg,
	}
}

// BgBrightRed background color (101).
func (a *Aurora) BgBrightRed(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgBrightRed()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgBrightRed()),
		value: arg,
	}
}

// BgBrightGreen background color (102).
func (a *Aurora) BgBrightGreen(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgBrightGreen()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgBrightGreen()),
		value: arg,
	}
}

// BgBrightYellow background color (103).
func (a *Aurora) BgBrightYellow(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgBrightYellow()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgBrightYellow()),
		value: arg,
	}
}

// BgBrightBlue background color (104).
func (a *Aurora) BgBrightBlue(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgBrightBlue()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgBrightBlue()),
		value: arg,
	}
}

// BgBrightMagenta background color (105).
func (a *Aurora) BgBrightMagenta(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgBrightMagenta()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgBrightMagenta()),
		value: arg,
	}
}

// BgBrightCyan background color (106).
func (a *Aurora) BgBrightCyan(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgBrightCyan()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgBrightCyan()),
		value: arg,
	}
}

// BgBrightWhite background color (107).
func (a *Aurora) BgBrightWhite(arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgBrightWhite()
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgBrightWhite()),
		value: arg,
	}
}

// Other background colors.
//
// BgIndex of 8-bit pre-defined background color from 0 to 255 (48;5;n).
//
//	  0-  7:  standard colors (as in ESC [ 40–47 m)
//	  8- 15:  high intensity colors (as in ESC [100–107 m)
//	 16-231:  6 × 6 × 6 cube (216 colors): 16 + 36 × r + 6 × g + b (0 ≤ r, g, b ≤ 5)
//	232-255:  grayscale from black to white in 24 steps
func (a *Aurora) BgIndex(n ColorIndex, arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgIndex(n)
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgIndex(n)),
		value: arg,
	}
}

// BgGray from 0 to 23.
func (a *Aurora) BgGray(n GrayIndex, arg interface{}) Value {
	if val, ok := arg.(Value); ok {
		return val.BgGray(n)
	}
	return Value{
		cc:    a.cc | colorConfig(Color(0).BgGray(n)),
		value: arg,
	}
}

// Special color functions.
//
// Colorize removes existing colors and
// formats of the argument and applies given.
func (a *Aurora) Colorize(arg interface{}, color Color) Value {
	if val, ok := arg.(Value); ok {
		return val.Colorize(color)
	}
	return Value{
		cc:    a.cc | colorConfig(color),
		value: arg,
	}
}

// Hyperlinks feature
//
// Hyperlink with given target and parameters. If hyperlinks feature is
// disabled, then the 'arg' argument dropped and the 'target' used instead
// inheriting all colors and format from the 'arg' (if it's a Colored).
//
// See https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda
// for details about the hyperlinks feature.
//
// The Hyperlink doesn't escape the target and the params. They should be
// checked and escaped before.
//
// See also HyperlinkID function.
//
// For a simple example
//
//	au.Hyperlink("Example", "http://example.com")
//
// and an example with ID
//
//	au.Hyperlink("Example", "http://example.com", aurora.HyperlinkID("10"))
func (a *Aurora) Hyperlink(arg interface{}, target string,
	params ...HyperlinkParam) Value {

	if val, ok := arg.(Value); ok {
		return val.Hyperlink(target, params...)
	}
	return Value{
		cc:    a.cc,
		value: arg,
	}.Hyperlink(target, params...)
}

// HyperlinkTarget of the argument if it's a Value.
func (a *Aurora) HyperlinkTarget(arg interface{}) (target string) {
	if val, ok := arg.(Value); ok {
		return val.HyperlinkTarget()
	}
	return // no target
}

// HyperlinkParams of the argument if it's a Value.
func (a *Aurora) HyperlinkParams(arg interface{}) (params []HyperlinkParam) {
	if val, ok := arg.(Value); ok {
		return val.HyperlinkParams()
	}
	return // no target
}

func (a *Aurora) transform(arg interface{}) (val Value, ok bool) {
	var ai Value
	ai, ok = arg.(Value)
	if !ok {
		return // Value{}, false
	}
	// if ai.cc.resetColor() == a.cc.resetColor() {
	// 	return // don't replace, same configurations
	// }
	val = Value{cc: a.cc | colorConfig(ai.cc.color()), value: ai.value}
	if a.cc.hyperlinksEnbaled() {
		val.hyperlink = ai.hyperlink
	}
	return val, true // transformed value, true
}

// Sprintf allows to use Value as format. For example
//
//	var v = Sprintf(Red("total: +3.5f points"), Blue(3.14))
//
// In this case "total:" and "points" will be red, but
// 3.14 will be blue. But, in another example
//
//	var v = Sprintf(Red("total: +3.5f points"), 3.14)
//
// full string will be red. And no way to clear 3.14 to default format and
// color.
//
// It applies own configurations to all given Values.
func (a *Aurora) Sprintf(format interface{}, args ...interface{}) string {
	// // clear colors & links as configured by the a
	if f, ok := a.transform(format); ok {
		format = f
	}
	for i := range args {
		if ax, ok := a.transform(args[i]); ok {
			args[i] = ax
		}
	}
	return sprintf(format, args...)
}
